pkg/auth/oauth2_client.go

```diff
@@ -0,0 +1,191 @@
+package auth
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/flyteorg/flytestdlib/logger"
+	goauth "golang.org/x/oauth2"
+	"google.golang.org/grpc"
+	"math/big"
+	"net/http"
+	"os"
+	"os/exec"
+	"time"
+)
+
+// The following provides the setup required for the client to perform the "Authorization Code" flow with PKCE in order
+// to obtain an access token for public/untrusted clients.
+
+const cookiePKCE = "isPKCE"
+
+var (
+	// pkceCodeVerifier stores the generated random value which the client will on-send to the auth server with the received
+	// authorization code. This way the oauth server can verify that the base64URLEncoded(sha265(codeVerifier)) matches
+	// the stored code challenge, which was initially sent through with the code+PKCE authorization request to ensure
+	// that this is the original user-agent who requested the access token.
+	PkceCodeVerifier string
+
+	// pkceCodeChallenge stores the base64(sha256(codeVerifier)) which is sent from the
+	// client to the auth server as required for PKCE.
+	PkceCodeChallenge string
+
+
+	TokenChannel chan string
+)
+
+// The following sets up the requirements for generating a standards compliant PKCE code verifier.
+const codeVerifierLenMin = 43
+const codeVerifierLenMax = 128
+const codeVerifierAllowedLetters = "abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ._~"
+
+// generateCodeVerifier provides an easy way to generate an n-length randomised
+// code verifier.
+func generateCodeVerifier(n int) string {
+	// Enforce standards compliance...
+	if n < codeVerifierLenMin {
+		n = codeVerifierLenMin
+	}
+	if n > codeVerifierLenMax {
+		n = codeVerifierLenMax
+	}
+
+	// Randomly choose some allowed characters...
+	b := make([]byte, n)
+	for i := range b {
+		// ensure we use non-deterministic random ints.
+		j, _ := rand.Int(rand.Reader, big.NewInt(int64(len(codeVerifierAllowedLetters))))
+		b[i] = codeVerifierAllowedLetters[j.Int64()]
+	}
+
+	return string(b)
+}
+
+// generateCodeChallenge returns a standards compliant PKCE S(HA)256 code
+// challenge.
+func generateCodeChallenge(codeVerifier string) string {
+	// Create a sha-265 hash from the code verifier...
+	s256 := sha256.New()
+	s256.Write([]byte(codeVerifier))
+
+	// Then base64 encode the hash sum to create a code challenge...
+	return base64.RawURLEncoding.EncodeToString(s256.Sum(nil))
+}
+
+// resetPKCE cleans up PKCE details and returns the code verifier.
+func resetPKCE() (codeVerifier string) {
+	codeVerifier = PkceCodeVerifier
+	PkceCodeVerifier = ""
+	return codeVerifier
+}
+
+
+// A valid oauth2 client (check the store) that additionally requests an OpenID Connect id token
+var clientConf = goauth.Config{
+	ClientID:     "flytectl",
+	ClientSecret: "foobar",
+	RedirectURL:  "http://localhost:3846/callback",
+	Scopes:       []string{"all"},
+	Endpoint: goauth.Endpoint{
```

Comment:
We should use the metadata URL to discover these... the reason being, when admin is configured to use Okta as an external authorization server, these endpoints will point to okta instead... the only way for the client to know that is if it always fetches (and caches?) `<adminUrl>/.well-known/oauth-authorization-server`

Reply:
Done.